src/build.cc

```diff
@@ -920,6 +912,20 @@ bool Builder::StartEdge(Edge* edge, string* err) {
       return false;
   }
 
+  // Find the most recent mtime of any (existing) non-order-only input
+  Node* most_recent_input = NULL;
+  for (vector<Node*>::iterator i = edge->inputs_.begin();
+        i != edge->inputs_.end() - edge->order_only_deps_; ++i) {
+    if (!(*i)->Stat(disk_interface_, err))
+      return false;
```

Comment:
Previously this extra `stat()` call was made only on `restat` edge inputs in `FinishCommand`'s `node_cleaned` case.  Now we're calling `stat()` a second time on every input of every edge that runs.  This may not matter because in practice the `stat()` result will probably be in the filesystem's hot cache, and the time it takes is likely small compared to the time it will take the edge's command to run (whose child process will normally actually open and read the whole input file).

I wonder if we can get away with `StatIfNecessary` here.  Even if the input has been updated since our original stat round when ninja started, we still know that the outputs will not be older than that.

Reply:
I thought hard about using `StatIfNecessary` here for awhile. I opted to re-stat because input nodes might be on multiple edges and later edges might be reading a different version of the input if that input is changed after one edge runs but before another begins.

This would cause a false positive on the subsequent run since the recorded mtime would reflect that input's original mtime even though the edge actually ran with the most recent version.

I suppose it's a tradeoff between saving unnecessary rebuilds on subsequent runs vs stat'ing more often while building.

Reply:
I thought hard about using StatIfNecessary here for awhile. I opted to re-stat because input nodes might be on multiple edges and later edges might be reading a different version of the input if that input is changed after one edge runs but before another begins.

This would cause a false positive on the subsequent run since the recorded mtime would reflect that input's original mtime even though the edge actually ran with the most recent version.

I suppose it's a tradeoff between saving unnecessary rebuilds on subsequent runs vs stat'ing more often while building.

Reply:
Thanks for explaining.  If a performance problem surfaces we can re-consider.  Otherwise I think `Stat()` is fine for now.